src/v/v8_engine/executor.h

```diff
@@ -0,0 +1,210 @@
+
+    static constexpr ss::lowres_clock::duration _max_wait_time_ms{100};
+
+    // _abort_sources is used for abort sleep if semaphore is stoped
+    ss::sharded<local_abort_source> _abort_sources;
+    std::atomic<bool> _is_closed{false};
+};
+
+} // namespace internal
+
+// This class implement thread pool with std::thread for runing v8 script. It
+// uses boost::lockfree_queue for storing task. Queue task stores #thread_count
+// tasks. Because we need to set watchdog for timeout v8 script only when
+// executor can execute script without wating. semaphore is used for controle
+// does executor have free thread.
+class executor {
+public:
+    executor(size_t threads_num, uint8_t cpu_id);
+
+    executor(const executor& other) = delete;
+    executor& operator=(const executor& other) = delete;
+    executor(executor&& other) = delete;
+    executor& operator=(executor&& other) = delete;
+
+    ~executor();
+
+    // Start ss::sharded for ss::gate and ss::abort_source
+    ss::future<> start();
+
+    // Stop executor. Frst of all we need to stop semaphore for cancel waitnig
+    // tasks. After that we need complete all task in queue and finally stop
+    // std::threads
+    ss::future<> stop();
+
+    bool is_stopping() const;
+
+    /// Submit new task in executor.
+    ///
+    /// \param start_func is used for set watchdog
+    /// \param func_for_executor is v8 script
+    template<typename StartFunc, typename FuncForExecutor>
+    auto submit(StartFunc&& start_func, FuncForExecutor&& func_for_executor) {
+        return ss::with_gate(
+          _gates.local()._gate,
+          [this,
+           start_func = std::move(start_func),
+           func_for_executor = std::move(func_for_executor)]() mutable {
+              return _sem.acquire()
+                .then(
+                  [this,
+                   start_func = std::move(start_func),
+                   func_for_executor = std::move(func_for_executor)]() mutable {
+                      auto new_task
+                        = std::make_unique<internal::task<FuncForExecutor>>(
+                          std::move(func_for_executor));
+                      auto fut = new_task->get_future();
+                      _tasks.push(new_task.get());
+                      _has_task_cv.notify_one();
+                      start_func(); // Rearm watchdog for srcipt
+                      return fut.finally([new_task = std::move(new_task)] {});
+                  })
+                .finally([this] { _sem.release(); });
+          });
+    }
+
+private:
+    // Do we need to pin all thread for one core?
+    void pin(unsigned cpu_id);
+
+    // Main loop for threads in executor
+    void loop();
+
+    std::atomic<bool> _is_stopped{false};
+
+    std::mutex mutex;
+    std::condition_variable _has_task_cv;
+    std::vector<std::thread> _threads; // maybe use ss::alien_thread
+
+    ss::sharded<internal::local_gate> _gates;
+    internal::semaphore _sem;
```

Comment:
I see two patterns here:
1. use of sharded (`_gates`)
2. use of a non-sharded class that has atomics within it (`_sem`)

We should pick one abstraction or another its kind of confusing at times!

Reply:
Yes, I think I can use gate inside semaphore

Reply:
I was thinking just don't use the seastar API at all within the executor. Just use the std c++ library primitives, so you can use the std::atomic<bool> to act like a gate, instead of a gate on each core

Reply:
I think I can delete gate from executor and used it inside semaphore with source_abort
Is it better?

Reply:
It is really cool things for stoping current execution in futures and throw exception

Reply:
Done. I did not use gates in executor